### PR TITLE
Add GitHub tests for karmadactl init --config command

### PR DESCRIPTION
**Pull request by tiansuo114**
**What type of PR is this?**

**What this PR does / why we need it**:
This PR is an additional test, based on the suggestion provided in [url](https://github.com/karmada-io/karmada/pull/5357#discussion_r1807568543).
**Which issue(s) this PR fixes**:
Part of #5387
It adds test cases for deploying Karmada services using the `karmadactl init --config` command with a configuration file, similar to the existing tests for `karmadactl init`.

**Special notes for your reviewer**:
@liangyuanpeng 
@RainbowMango 

~~This PR currently contains two commits because the test files depend on the new feature for deploying with a configuration file, introduced in [#5357](https://github.com/karmada-io/karmada/pull/5357). Once that PR is merged, this PR will be rebased, retaining only the commit for adding test cases. However, to save time, it is being submitted for review in advance.~~

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

**Comment by tiansuo114** [comment]
May need help😭😭, in my own project's github action, these tests passed correctly, but when I commit, these tests become failed, what should I do to fix these problems
@liangyuanpeng

File: pkg/karmadactl/cmdinit/kubernetes/deploy.go
Comment (liangyuanpeng) [inline pkg/karmadactl/cmdinit/kubernetes/deploy.go:739]: does we have not constants for it?

File: pkg/karmadactl/config/init/list_option.go
Comment (liangyuanpeng) [inline pkg/karmadactl/config/init/list_option.go:45]: Seems like this comment is nocorrect.

File: pkg/karmadactl/config/init/list_option.go
Comment (liangyuanpeng) [inline pkg/karmadactl/config/init/list_option.go:26]:
```suggestion
// CommandConfigImageOption options for components.
```

**Review by liangyuanpeng — COMMENTED** [review]
Thanks you working for it, could you please add some e2e test here?  Maybe take a look at how kubeadm does it.

File: pkg/karmadactl/cmdinit/kubernetes/deploy.go
Comment (tiansuo114) [reply pkg/karmadactl/cmdinit/kubernetes/deploy.go:739]: I was not observed in the search system, which has the corresponding constants, in NewDefaultCommandInitOption approach is to write, I mainly refer to is the PKG/karmadactl/cmdinit/cmdinit. Go in for the default value of cmdflag, there are Parts of const variables are mostly copied using const constants

Will need these for example KubeImageTag string constant changes in the form of variables on the PKG/karmadactl/cmdinit/kubernets/deploy. Go in?

**Comment by tiansuo114** [comment]
> Thanks you working for it, could you please add some e2e test here? Maybe take a look at how kubeadm does it.

I have fixed some bugs and added e2e tests, please let me know if you need to make further changes

**Comment by tiansuo114** [comment]
> Thanks you working for it, could you please add some e2e test here? Maybe take a look at how kubeadm does it.

I have a question regarding the construction of the karmadactl application in the E2E test environment. While observing the workflow in the setup e2e test environment section, I did not find any actions to compile it using the Makefile. Without using the Makefile for compilation, the -X github.com/karmada-io/karmada/pkg/version.gitVersion=v1.11.0-alpha.0-254-gf3bcb21ff-dirty linker flag is not applied, which might be the reason why the output of karmadactl config image list in the E2E test is different from my expected output. How should I solve this issue?

**Comment by tiansuo114** [comment]
> Thanks you working for it, could you please add some e2e test here? Maybe take a look at how kubeadm does it.感谢您的工作，您可以在此处添加一些 e2e 测试吗？也许看看 kubeadm 是如何做到的。

"I have fixed the issues you pointed out and provided new e2e test cases. Please let me know if there are any further modifications needed. :)

**Comment by liangyuanpeng** [comment]
> This PR is related to the documentation PR: https://github.com/karmada-io/karmada/pull/5277

@tiansuo114 

Add to PR issue body would be great, thank you! 

image

**Comment by tiansuo114** [comment]
> > This PR is related to the documentation PR: #5277此 PR 与文档 PR 相关： #5277
> 
> @tiansuo114
> 
> Add to PR issue body would be great, thank you!添加到 PR 问题正文会很棒，谢谢！
> 
> image

Ok, I modified the pr and the other two questions about pr and added links to it

**Comment by XiShanYongYe-Chang** [comment]
Ask @zhzhuang-zju have a look.
/assign @zhzhuang-zju

File: pkg/karmadactl/karmadactl.go
Comment (zhzhuang-zju) [inline pkg/karmadactl/karmadactl.go:100]: The `config` command serves the `init` command, so it would be better to place it in the "Cluster Registration Commands:" group.

File: pkg/karmadactl/cmdinit/kubernetes/deploy.go
Comment (zhzhuang-zju) [reply pkg/karmadactl/cmdinit/kubernetes/deploy.go:739]: > Will need these for example KubeImageTag string constant changes in the form of variables on the PKG/karmadactl/cmdinit/kubernets/deploy. Go in?

How about creating constants like `KubeImageTag` and `KarmadaNamespace` to reduce shotgun surgery modifications?

File: pkg/karmadactl/cmdinit/kubernetes/deploy.go
Comment (zhzhuang-zju) [inline pkg/karmadactl/cmdinit/kubernetes/deploy.go:760]: func `NewDefaultCommandInitOption` is intended for use with the `config` command, but the `config` command actually only uses a few of the options. Is it necessary to set default values for all options?

Comment (tiansuo114) [reply pkg/karmadactl/cmdinit/kubernetes/deploy.go:760]: I think it could be used as a general initialization operation to facilitate the addition of other commands related to karmadactl config in the future. Adding all the default information here might be helpful for future development work, but removing this information and adding it only when needed is also an option. Which approach do you think is better?

File: pkg/karmadactl/cmdinit/kubernetes/deploy.go
Comment (tiansuo114) [reply pkg/karmadactl/cmdinit/kubernetes/deploy.go:739]: > > Will need these for example KubeImageTag string constant changes in the form of variables on the PKG/karmadactl/cmdinit/kubernets/deploy. Go in?将需要将这些例如 KubeImageTag 字符串常量以变量的形式放在 PKG/karmadactl/cmdinit/kubernets/deploy 上。进去？
> 
> How about creating constants like `KubeImageTag` and `KarmadaNamespace` to reduce shotgun surgery modifications?创建像 `KubeImageTag` 和 `KarmadaNamespace` 这样的常量来减少霰弹枪手术修改怎么样？

Ok, I'll add some const variables to the config package and modify the code

File: pkg/karmadactl/cmdinit/kubernetes/deploy.go
Comment (zhzhuang-zju) [reply pkg/karmadactl/cmdinit/kubernetes/deploy.go:760]: > Adding all the default information here might be helpful for future development work, but removing this information and adding it only when needed is also an option. Which approach do you think is better?

I perfer the latter. If subsequent development requires a complete default initialization operation, we can extract the operations from `flags.XXXVarP` to obtain a generic initialization operation that includes default values.

Comment (tiansuo114) [reply pkg/karmadactl/cmdinit/kubernetes/deploy.go]: > > Adding all the default information here might be helpful for future development work, but removing this information and adding it only when needed is also an option. Which approach do you think is better?
> 
> I perfer the latter. If subsequent development requires a complete default initialization operation, we can extract the operations from `flags.XXXVarP` to obtain a generic initialization operation that includes default values.

I have fixed the defects mentioned above, how does the code look now

**Comment by liangyuanpeng** [comment]
/hold 
need squash the commits when ready to merge.

**Comment by tiansuo114** [comment]
> /hold /抓住 need squash the commits when ready to merge.准备合并时需要压缩提交。

get,Ok, I will squash commit after the problem in comment is solved

**Comment by tiansuo114** [comment]
> > /hold /抓住 need squash the commits when ready to merge.准备合并时需要压缩提交。
> 
> get,Ok, I will squash commit after the problem in comment is solvedget，好的，评论中的问题解决后我会squash commit

File: pkg/karmadactl/config/image.go
Comment (tiansuo114) [reply pkg/karmadactl/config/image.go]: Now that the comments have been fixed, how does the code look now

File: pkg/karmadactl/config/list.go
Comment (zhzhuang-zju) [inline pkg/karmadactl/config/list.go:38]: Could we consider removing the function `NewDefaultCommandInitOption`, as default values can all be set through `cmd.Flags().XXXVarP`?
And if testing requires it, we could also move `NewDefaultCommandInitOption` to the corresponding test file.

File: pkg/karmadactl/config/list.go
Comment (zhzhuang-zju) [reply pkg/karmadactl/config/list.go:31]: @tiansuo114 The original sentence is not very smooth; it would be better if it could be modified~

File: pkg/karmadactl/config/list.go
Comment (liangyuanpeng) [reply pkg/karmadactl/config/list.go:38]: @tiansuo114  could you explain why we need it?

Comment (tiansuo114) [reply pkg/karmadactl/config/list.go:38]: > Could we consider removing the function `NewDefaultCommandInitOption`, as default values can all be set through `cmd.Flags().XXXVarP`?我们是否可以考虑删除函数 `NewDefaultCommandInitOption` ，因为默认值都可以通过 `cmd.Flags().XXXVarP` 设置？ And if testing requires it, we could also move `NewDefaultCommandInitOption` to the corresponding test file. 如果测试需要，我们也可以将 `NewDefaultCommandInitOption` 移动到相应的测试文件中。

I am attempting to discard the NewDefaultCommandInitOption method and use the following method to create CommandInitOption in the file.
```
func NewCmdConfigImagesList(parentCommand string) *cobra.Command {
	opts := configInit.CommandConfigImageOption{
		InitOption: &kubernetes.CommandInitOption{},
	}
	cmd := &cobra.Command{
```
However, when I actually execute it, the command produces the following error.
```
registry.k8s.io/kube-apiserver:
registry.k8s.io/kube-controller-manager:
registry.k8s.io/etcd:3.5.13-0



```
When the NewDefaultCommandInitOption initialization is not performed, the output string related to the image will be empty.
I haven't found a good way to solve this issue, so I have continued to keep the NewDefaultCommandInitOption method in use.

Comment (zhzhuang-zju) [reply pkg/karmadactl/config/list.go:38]: So, the `karmadactl config image list` command currently does not support configuring image tags, correct?

Comment (tiansuo114) [reply pkg/karmadactl/config/list.go:38]: In what is now the code is to support the image of the tag, the image of the tags related to fill github.com/karmada-io/karmada/pkg/version.gitVersion fields at compile time

Comment (tiansuo114) [reply pkg/karmadactl/config/list.go:38]: @zhzhuang-zju 
I have added the output that the command would normally provide at the beginning of the PR. Regarding the question in the comment about the NewDefaultCommandInitOption method, what do you think would be the most appropriate way to handle this? Should we keep the currently working method, or do you have a better solution to suggest?

Comment (liangyuanpeng) [reply pkg/karmadactl/config/list.go:38]: > So, the command currently does not support configuring image tags, correct?karmadactl config image list

how about let's reduce review pressure to quickly move forward with PR consolidation, leave it to the next PR.

I think you're talking about the flag of `kube-image-tag` (we have not karmada-imag-tag yet)

File: test/e2e/karmadactl_test.go
Comment (liangyuanpeng) [inline test/e2e/karmadactl_test.go:1272]: Once you rebase PR, this will need to be 1.30.4 and not 1.29.6

File: pkg/karmadactl/cmdinit/options/global.go
Comment (liangyuanpeng) [inline pkg/karmadactl/cmdinit/options/global.go:46]:
```suggestion
	DefaultKubeImageTag = "v1.30.4"
```

**Comment by liangyuanpeng** [comment]
@zhzhuang-zju  anyelse comment here?

File: pkg/karmadactl/config/list.go
Comment (zhzhuang-zju) [reply pkg/karmadactl/config/list.go:38]: > I think you're talking about the flag of kube-image-tag (we have not karmada-imag-tag yet)

Yes, I think it would be much better to introduce the flag `kube-image-tag`. 
@tiansuo114  Could you help confirm whether the function `NewDefaultCommandInitOption` will still be needed after introducing the flag `kube-image-tag`? Thanks.

Comment (tiansuo114) [reply pkg/karmadactl/config/list.go:38]: @zhzhuang-zju 
In my understanding, the output of the config image list command consists of three parts:
repository link + image name + image tag
In this command, the first two parts rely on the initialization of kubernetes.CommandInitOption. Therefore, even after introducing kube-image-tag, the NewDefaultCommandInitOption method is still necessary.

Comment (zhzhuang-zju) [reply pkg/karmadactl/config/list.go:38]: Thank you for your answer. If there are no current plans to expose more configurable flags, the current implementation is fine with me.

Comment (tiansuo114) [reply pkg/karmadactl/config/list.go:38]: get,I will fix the remaining issues as soon as possible and prepare for the next step

File: pkg/karmadactl/cmdinit/options/global.go
Comment (tiansuo114) [reply pkg/karmadactl/cmdinit/options/global.go]: have fixed and rebased

**Comment by tiansuo114** [comment]
> @zhzhuang-zju anyelse comment here?

Can this pr be merged now or does it need to be further modified😭😭

**Comment by zhzhuang-zju** [comment]
/lgtm
assign @XiShanYongYe-Chang

**Comment by tiansuo114** [comment]
The command in this feature PR has been modified to match the [current documentation](https://github.com/karmada-io/karmada/pull/5623) format, but its functionality remains unchanged and is still the same as it was initially.
@liangyuanpeng @RainbowMango

**Comment by tiansuo114** [comment]
Based on community meetings, it seems that this feature is not necessary and will increase maintenance costs, so this pr is temporarily closed

**Comment by liangyuanpeng** [comment]
https://github.com/karmada-io/karmada/pull/5744 is migration to here (for Summer OSPP 2024)

**Comment by tiansuo114** [comment]
Based on the OSPP project's requirements, this PR has been reopened. However, the relevant content in this PR has been changed to add automated testing for the new `karmadactl init --config` command as indicated in the [link](https://github.com/karmada-io/karmada/pull/5744).

@liangyuanpeng @RainbowMango @zhzhuang-zju

**Comment by tiansuo114** [comment]
> #5744 is migration to here (for Summer OSPP 2024)

The code in this PR has been re-pushed and tested, and after a code clean-up, I’ve removed the unnecessary metadata part in KarmadaInitConfig. We can now continue with the previous review.🎉

If need to track this [issue](https://github.com/karmada-io/karmada/pull/5357#discussion_r1818713706), please refer to the relevant changes in this PR. 
@RainbowMango @liangyuanpeng

**Review by liangyuanpeng — COMMENTED** [review]
thank you!
/lgtm 
/cc @zhzhuang-zju  if you want to review for karmadactl

File: .github/workflows/cli.yaml
Comment (RainbowMango) [inline .github/workflows/cli.yaml:106]: end of line.

**Review by RainbowMango — COMMENTED** [review]
Looks good to me, just a nit, otherwise LGTM.

File: .github/workflows/cli.yaml
Comment (tiansuo114) [reply .github/workflows/cli.yaml]: Do you mean it’s best to add a blank line at the end of the line? If so, the code has already been updated. If I’ve misunderstood, please feel free to correct me.

Comment (liangyuanpeng) [reply .github/workflows/cli.yaml]: > Do you mean it’s best to add a blank line at the end of the line? 

correct

**Review by RainbowMango — COMMENTED** [review]
/lgtm
/assign @liangyuanpeng

**Comment by RainbowMango** [comment]
It seems this is the last PR for the summer OSPP program. #5387.
Please let me know if I miss anything.

**Comment by tiansuo114** [comment]
> It seems this is the last PR for the summer OSPP program. #5387. Please let me know if I miss anything.


Yes, this PR is the last one for the OSPP tasks. Up until now, with the merge of this PR into the main branch, all planned PRs have been incorporated into Karmada.
PR Link 1: https://github.com/karmada-io/karmada/pull/5277
PR Link 2: https://github.com/karmada-io/karmada/pull/5306
PR Link 3: https://github.com/karmada-io/karmada/pull/5357
PR Link 4: https://github.com/karmada-io/website/pull/690
PR Link 5: https://github.com/karmada-io/karmada/pull/5321

**Comment by liangyuanpeng** [comment]
> It seems this is the last PR for the summer OSPP program. https://github.com/karmada-io/karmada/issues/5387.

Yes, it is.
/lgtm
/approve

**Review by RainbowMango — APPROVED** [review]
/approve

**Pull request merged**
